controllers/openstackipset_controller.go

```diff
@@ -246,6 +245,199 @@ func (r *OpenStackIPSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
+		}
+
+		// update status of OSNet
+		err = r.Client.Status().Update(context.TODO(), network)
+		if err != nil {
+			r.Log.Error(err, "Failed to update OpenStackNet status %v")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *OpenStackIPSetReconciler) addIP(instance *ospdirectorv1beta1.OpenStackIPSet, newHosts common.HostnameList) error {
+
+	for _, k := range newHosts {
+		//hostRef := k.HostRef
+		hostname := k.Hostname
+
+		hostNetworkIPs := map[string]string{}
+
+		// iterate over the requested Networks
+		for _, netName := range instance.Spec.Networks {
+
+			network := &ospdirectorv1beta1.OpenStackNet{}
+			err := r.Client.Get(context.TODO(), types.NamespacedName{Name: netName, Namespace: instance.Namespace}, network)
+			if err != nil {
+				if k8s_errors.IsNotFound(err) {
+					r.Log.Info(fmt.Sprintf("OpenStackNet named %s not found!", netName))
+				}
+				// Error reading the object - requeue the request.
+				return err
+			}
+
+			// set ipset NetworkStatus
+			instance.Status.Networks[network.Name] = ospdirectorv1beta1.NetworkStatus{
+				Cidr:            network.Spec.Cidr,
+				Vlan:            network.Spec.Vlan,
+				AllocationStart: network.Spec.AllocationStart,
+				AllocationEnd:   network.Spec.AllocationEnd,
+				Gateway:         network.Spec.Gateway,
+			}
+
+			_, cidr, err := net.ParseCIDR(network.Spec.Cidr)
+			if err != nil {
+				r.Log.Error(err, fmt.Sprintf("Failed to parse CIDR %s", network.Spec.Cidr))
+				return err
+			}
+
+			r.Log.Info(fmt.Sprintf("Network %s, cidr %v", network.Name, cidr))
+
+			start := net.ParseIP(network.Spec.AllocationStart)
+			end := net.ParseIP(network.Spec.AllocationEnd)
+
+			reservationIP := ""
+
+			// Do we already have a reservation for this hostname on the network?
+			for _, reservation := range network.Status.RoleReservations[instance.Spec.RoleName].Reservations {
+				if reservation.Hostname == hostname {
+					// We also need the netmask (which is not stored in the OpenStackNet status),
+					// so we acquire it from the OpenStackIPSet spec
+					cidrPieces := strings.Split(network.Spec.Cidr, "/")
+					reservationIP = fmt.Sprintf("%s/%s", reservation.IP, cidrPieces[len(cidrPieces)-1])
+					break
+				}
+			}
+
+			// get list of all reservations
+			reservationList := []ospdirectorv1beta1.IPReservation{}
+			for _, roleReservations := range network.Status.RoleReservations {
+				reservationList = append(reservationList, roleReservations.Reservations...)
+			}
+
+			if reservationIP == "" {
+				r.Log.Info(fmt.Sprintf("Create new reservation for host %s in network %s", hostname, netName))
+
+				// No reservation found, so create a new one
+				ip, reservation, err := common.AssignIP(common.AssignIPDetails{
+					IPnet:           *cidr,
+					RangeStart:      start,
+					RangeEnd:        end,
+					RoleReservelist: network.Status.RoleReservations[instance.Spec.RoleName].Reservations,
```

Comment:
Is it possible that `network.Status.RoleReservations[instance.Spec.RoleName]` could point to a `nil` or zero-value entry?  If so, maybe that could cause a problem when it then tries to access non-existent `.Reservations`.

Reply:
change to initialize them if it does not exist